### ISSUE TITLE
Provide an option to check for newer terraform plugin versions 

### ISSUE THREAD

**Issue by adamdecaf** [hashicorp/terraform#16211]
It would be nice if terraform could check and show the updated versions from the plugins release service. Since [semver](http://semver.org/) is being used it should be possible to also show major/minor/bugfix upgrades possible (if the `~>` flag isn't used). 

There are similar tools in other languages ecosystems that I've liked to use which could be a basis for work in terraform. In one such tool (sbt-updates) a command is ran (`dependencyUpdates`) which crawls the artifact repository and shows bugfix, minor, and major updates available. 

```
[info]   com.warrenstrange:googleauth                         : 1.1.1    -> 1.1.2
[info]   com.banno:redis-utils                                : 4.13.2   -> 4.13.3   -> 4.17.2    -> 7.6.1
[info]   com.banno:scala-vault-api                            : 3.6.2                             -> 4.0.0
```

Related work: https://github.com/rtimush/sbt-updates

**Comment by apparentlymart** [comment]
Hi @adamdecaf! Thanks for sharing this idea.

I think what you want to see here is whether there are new versions available that meet the specified version constraints. Is that right? In other words, if I've set `version = "~> 1.0"` then I might see that `1.0.1` is available, but not `1.1.0`.

We've been considering turning `terraform providers` into a subcommand family so we could have other commands in there to enable things like surgically upgrading just one provider. If we go that route, it seems reasonable that this feature would live there. For example:

```
$ terraform providers show-upgrades
Provider    Current  Available  Newest
aws         1.0.1    1.0.2      3.0.0
template    1.0.0    1.0.1      1.0.1
```

I suppose this would show only providers where either "Available" (newest version meeting constraint) or "Newest" (newest version ignoring constraints) are different than "Current".

The idea of showing all of the different newer versions for each of a patch, minor or major upgrade is an interesting one. The example shown here seems confusing though, and I'm not sure I would've understood what it was telling me had you not explained it. I wonder if showing just the two other versions in the above example deals with the 80% case: I can see both whether there's something available within my constraint and if there's possibly-more-risky upgrade available, and then use the changelog to evaluate the impact of the latter.

**Comment by adamdecaf** [comment]
Yep! The example project above doesn't include formatting headers, which your example does, however it is showing the furthest "jumps" available for bugfix, minor and major version upgrades for each dependency. 

I think showing `Available` and `Newest` will work for the 80% case.

**Comment by sumanmaity1234** [comment]
Hi @apparentlymart, any update on this?

**Comment by apparentlymart** [comment]
As far as I know, nobody has completed any further design work to figure out how a feature like this might work.

However, one thing that _has_ changed in the mean time is that Terraform now tracks provider dependency selections in [the dependency lock file](https://www.terraform.io/docs/language/dependency-lock.html). Although this is not a tailored solution to the particular use-case, one possible way to get the same information would be:

* Run `terraform init -upgrade` to change the selections to the newest matching the version constraints in the configuration.
* Use your version control system to inspect the diff of the dependency lock file, to see which new versions Terraform selected for each provider.
* If you want to keep those new selections then commit them to version control. Otherwise, discard the changes to return to the previous lock file contents.

I share this only in the hope that it might be useful as an interim solution. The other thing that changed since this original discussion is that GitHub's dependabot has support for proposing Terraform provider upgrades itself, and so I believe some folks are using that as a way to be notified (by a pull request, in that case) about new provider releases.
